File: backend/jobplatform/src/main/java/edu/tlu/jobplatform/subscription/domain/model/CandidateSubscription.java
package edu.tlu.jobplatform.subscription.domain.model;

import lombok.Builder;
import lombok.Getter;

import java.time.LocalDateTime;
import java.util.UUID;

/**
 * Aggregate Root: Gói dịch vụ của Ứng viên (Candidate).
 *
 * Reset policy:
 * - applicationQuota → reset đầu mỗi tháng (scheduler)
 * - cvBoostQuota → reset đầu mỗi tháng
 * - cvCreateQuota → NOT reset — giới hạn số CV tồn tại đồng thời
 */
@Getter
@Builder
public class CandidateSubscription {

    private final UUID id;
    private final UUID candidateId;
    private final UUID planId;
    private final String planCode; // "FREE_CANDIDATE", "PRO", "PREMIUM"
    private final boolean yearly;

    private LocalDateTime startedAt;
    private LocalDateTime expiresAt;
    private CandidateSubscriptionStatus status;

    // ── Quota theo tháng (reset hàng tháng) ──────────────────────────

    /** Số đơn ứng tuyển / tháng. FREE_CANDIDATE=5, PRO/PREMIUM=-1 */
    private CandidateQuota applicationQuota;

    /** Số lần boost CV / tháng. FREE_CANDIDATE=0, PRO=3, PREMIUM=-1 */
    private CandidateQuota cvBoostQuota;

    // ── Quota tổng (không reset) ──────────────────────────────────────

    /**
     * Số CV online có thể tạo đồng thời.
     * FREE_CANDIDATE=1, PRO=5, PREMIUM=-1.
     * Không reset — dùng để check trước khi tạo CV mới.
     * Không consume/refund theo đơn — track bằng count DB thực tế.
     */
    private CandidateQuota cvCreateQuota;

    // ── Feature flags ─────────────────────────────────────────────────

    /** AI viết & tối ưu CV theo JD — chỉ PREMIUM */
    private final boolean aiCvWriter;

    /**
     * Được dùng template premium khi tạo CV online.
     * PRO và PREMIUM = true, FREE_CANDIDATE = false.
     */
    private final boolean premiumTemplateAccess;

    private final LocalDateTime createdAt;
    private UUID currentPaymentId;
    private LocalDateTime lastQuotaResetAt;

    // ── Business Rules ────────────────────────────────────────────────

    public boolean isActive() {
        if (status != CandidateSubscriptionStatus.ACTIVE)
            return false;
        if (expiresAt == null)
            return true; // FREE plan — vĩnh viễn
        return LocalDateTime.now().isBefore(expiresAt);
    }

    public boolean isExpired() {
        if (status == CandidateSubscriptionStatus.EXPIRED)
            return true;
        if (expiresAt == null)
            return false;
        return LocalDateTime.now().isAfter(expiresAt);
    }

    public long daysRemaining() {
        if (!isActive())
            return 0;
        if (expiresAt == null)
            return Long.MAX_VALUE;
        return java.time.Duration.between(LocalDateTime.now(), expiresAt).toDays();
    }

    public boolean canApply() {
        return isActive() && !applicationQuota.isExceeded();
    }

    public boolean canBoostCv() {
        return isActive() && !cvBoostQuota.isExceeded();
    }

    public boolean isFree() {
        return PlanCode.FREE_CANDIDATE.equalsIgnoreCase(planCode);
    }

    // ── State transitions ─────────────────────────────────────────────

    public void activate(UUID paymentId, LocalDateTime expiresAt) {
        this.status = CandidateSubscriptionStatus.ACTIVE;
        this.startedAt = LocalDateTime.now();
        this.expiresAt = expiresAt;
        this.currentPaymentId = paymentId;
        this.lastQuotaResetAt = LocalDateTime.now();
    }

    public void expire() {
        this.status = CandidateSubscriptionStatus.EXPIRED;
    }

    public void cancel() {
        this.status = CandidateSubscriptionStatus.CANCELLED;
    }

    public void markFailed() {
        this.status = CandidateSubscriptionStatus.FAILED;
    }

    // ── Quota operations ──────────────────────────────────────────────

    public void consumeApplication() {
        this.applicationQuota = applicationQuota.consume(1);
    }

    public void refundApplication() {
        this.applicationQuota = applicationQuota.refund(1);
    }

    public void consumeCvBoost() {
        this.cvBoostQuota = cvBoostQuota.consume(1);
    }

    /**
     * Reset quota hàng tháng.
     * cvCreateQuota KHÔNG reset — đây là giới hạn số CV đồng thời, không phải
     * lượt/tháng.
     */
    public void resetMonthlyQuotas() {
        this.applicationQuota = applicationQuota.reset();
        this.cvBoostQuota = cvBoostQuota.reset();
        this.lastQuotaResetAt = LocalDateTime.now();
    }
}

File: frontend/src/domain/models/CandidateSubscription.ts
// D:\TLU_JOB_PLATFORM\frontend\src\domain\models\CandidateSubscription.ts

export type SubscriptionStatus = 'ACTIVE' | 'EXPIRED' | 'CANCELLED' | 'PENDING' | 'FAILED';

export interface CandidateQuota {
  limit: number;
  used: number;
}

export interface CandidateSubscriptionPlan {
  id: string;
  code: string;
  name: string;
  description: string;
  priceMonthly: number;
  priceYearly: number;
  applicationLimit: number;
  cvBoostLimit: number;
  /** Số CV online có thể tạo đồng thời. -1 = unlimited */
  cvCreateLimit: number;
  /** AI viết & tối ưu CV theo JD — chỉ PREMIUM */
  aiCvWriter: boolean;
  /** Được dùng template premium khi tạo CV online */
  premiumTemplateAccess: boolean;
  durationDays: number;
  active: boolean;
  free: boolean;
}

export interface CandidateSubscription {
  id: string;
  candidateId: string;
  planId: string;
  planCode: string;
  yearly: boolean;
  startedAt: string;
  expiresAt: string;
  status: SubscriptionStatus;
  applicationQuota: CandidateQuota;
  cvBoostQuota: CandidateQuota;
  /** Quota tạo CV online */
  cvCreateQuota: CandidateQuota;
  aiCvWriter: boolean;
  premiumTemplateAccess: boolean;
  currentPaymentId?: string;
  lastQuotaResetAt?: string;
  createdAt: string;
}

export interface CandidateQuotaResult {
  applicationQuota: CandidateQuota;
  cvBoostQuota: CandidateQuota;
  cvCreateQuota: CandidateQuota;
}

export interface PurchaseRequest {
  planId: string;
  yearly: boolean;
}

export interface PurchaseResult {
  paymentUrl: string;
  orderId: string;
}

/** Payload cho POST /api/v1/admin/candidate-plans */
export interface CandidatePlanPayload {
  code: string;
  name: string;
  description?: string;
  /** null nếu là gói free */
  priceMonthly: number | null;
  priceYearly: number | null;
  /** Số đơn ứng tuyển / tháng. -1 = unlimited */
  applicationLimit: number;
  /** Số lần boost CV lên top / tháng. 0 = không có */
  cvBoostLimit: number;
  /** Số CV online có thể tạo đồng thời. -1 = unlimited */
  cvCreateLimit: number;
  /** AI viết & tối ưu CV theo JD */
  aiCvWriter: boolean;
  /** Được dùng template premium khi tạo CV online */
  premiumTemplateAccess: boolean;
  /** null nếu gói free không có thời hạn */
  durationDays: number | null;
  active: boolean;
  free: boolean;
}

/** Payload cho PATCH /api/v1/admin/candidate-plans/{planId} */
export interface CandidatePlanUpdatePayload {
  name?: string;
  description?: string;
  priceMonthly?: number;
  priceYearly?: number;
  applicationLimit?: number;
  cvBoostLimit?: number;
  cvCreateLimit?: number;
  aiCvWriter?: boolean;
  premiumTemplateAccess?: boolean;
  durationDays?: number;
  active?: boolean;
}

export interface CandidateSubscriptionRow {
  id: string;
  candidateId: string;
  candidateName: string;
  planCode: string;
  status: SubscriptionStatus;
  yearly: boolean;
  startedAt: string;
  expiresAt: string;
  amount: number;
}

export interface PageResult<T> {
  content: T[];
  totalElements: number;
  totalPages: number;
}